gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/impl/MessageServiceImpl.java

```diff
@@ -287,35 +287,11 @@ else if (api != null && RoleScope.APPLICATION.name().equals(recipientEntity.getR
                                 .map(MembershipEntity::getMemberId)
                                 .collect(Collectors.toSet())
                         );
-                        // get all indirect members
-                        applicationService
-                            .findByIds(context, applicationIds)
-                            .stream()
-                            .map(applicationEntity -> applicationEntity.getGroups())
-                            .reduce(
-                                (a, b) -> {
-                                    a.addAll(b);
-                                    return a;
-                                }
-                            )
-                            .ifPresent(
-                                applicationsGroups -> {
-                                    if (!applicationsGroups.isEmpty()) {
-                                        // get all indirect members
-                                        recipientIds.addAll(
-                                            membershipService
-                                                .getMembershipsByReferencesAndRole(
-                                                    MembershipReferenceType.GROUP,
-                                                    new ArrayList<>(applicationsGroups),
-                                                    optRole.get().getId()
-                                                )
-                                                .stream()
-                                                .map(MembershipEntity::getMemberId)
-                                                .collect(Collectors.toSet())
-                                        );
-                                    }
-                                }
-                            );
+
+                        Set<String> indirectMemberIds = this.getIndirectMemberIds(context, applicationIds, optRole.get());
+                        if (indirectMemberIds != null) {
```

Comment:
If we make `getIndirectMemberIds` returns an empty `Set` instead of `null` we can always do the `recipientIds.addAll` and so avoid this if